cpp/include/ray/api/ray_register.h

```diff
@@ -0,0 +1,201 @@
+template <class First, class... Second>
+struct AddType<First, std::tuple<Second...>> {
+  using type = std::tuple<First, Second...>;
+};
+
+template <class First, class Second>
+using AddType_t = typename AddType<First, Second>::type;
+
+template <typename OUT, typename IN>
+inline static OUT AddressOf(IN in) {
+  union {
+    IN in;
+    OUT out;
+  } u = {in};
+
+  return u.out;
+};
+
+class Router {
+ public:
+  static Router &Instance() {
+    static Router instance;
+    return instance;
+  }
+
+  template <typename Function>
+  bool RegisterHandler(std::string const &name, const Function &f) {
+    auto pair = func_ptr_to_key_map_.emplace(AddressOf<uintptr_t>(f), name);
+    if (!pair.second) {
+      return false;
+    }
+
+    RegisterNonmemberFunc(name, f);
+    return true;
+  }
+
+  std::string Route(const char *data, std::size_t size) {
+    std::string result;
+    do {
+      try {
+        auto p = Serializer::Deserialize<std::tuple<std::string>>(data, size);
+        auto &func_name = std::get<0>(p);
+        auto it = map_invokers_.find(func_name);
+        if (it == map_invokers_.end()) {
+          result = PackArgsStr(ErrorCode::FAIL, "unknown function: " + func_name);
+          break;
+        }
+
+        it->second(data, size, result);
+        if (result.size() >= MAX_BUF_LEN) {
+          result = PackArgsStr(
+              ErrorCode::FAIL,
+              "the response result is out of range: more than 10M " + func_name);
+        }
+      } catch (const std::exception &ex) {
+        result = PackArgsStr(ErrorCode::FAIL, ex.what());
+      }
+    } while (0);
+
+    return result;
+  }
+
+ private:
+  Router() = default;
+  ~Router() = default;
+  Router(const Router &) = delete;
+  Router(Router &&) = delete;
+
+  template <typename F, size_t... I, typename Arg, typename... Args>
+  static typename std::result_of<F(Args...)>::type CallHelper(
+      const F &f, const absl::index_sequence<I...> &, std::tuple<Arg, Args...> tup) {
+    return f(std::move(std::get<I + 1>(tup))...);
+  }
+
+  template <typename F, typename Arg, typename... Args>
+  static typename std::enable_if<
+      std::is_void<typename std::result_of<F(Args...)>::type>::value>::type
+  Call(const F &f, std::string &result, std::tuple<Arg, Args...> tp) {
+    CallHelper(f, absl::make_index_sequence<sizeof...(Args)>{}, std::move(tp));
+    result = PackArgsStr(ErrorCode::OK);
+  }
+
+  template <typename F, typename Arg, typename... Args>
+  static typename std::enable_if<
+      !std::is_void<typename std::result_of<F(Args...)>::type>::value>::type
+  Call(const F &f, std::string &result, std::tuple<Arg, Args...> tp) {
+    auto r = CallHelper(f, absl::make_index_sequence<sizeof...(Args)>{}, std::move(tp));
+    result = PackArgsStr(ErrorCode::OK, "ok", r);
+  }
+
+  template <typename Function>
+  struct invoker {
+    static inline void Apply(const Function &func, const char *data, size_t size,
+                             std::string &result) {
```

Comment:
Why `void Apply(func, data, size, result)` instead of `std::string Apply(func, data, size)`? BTW, for output parameters, you need to use pointers rather than references, according to Google C++ style.

Reply:
Good, return the result is more natural.